### PR TITLE
fix(tests) Add a compare test to the API tests

### PR DESCRIPTION
**Pull request by evanh**
Refactored the current SnQL test fixtures to add a 'compare' param that runs
both the SnQL and legacy query and compares the sql output. Added to most API
tests (test_api.py will be a separate PR because it's huge).

### CODE REVIEW

File: snuba/query/snql/parser.py [hunk 1]
```diff
@@ -118,7 +123,7 @@
     or_tuple              = space+ "OR" and_expression
 
     condition             = main_condition / parenthesized_cdn
-    main_condition        = low_pri_arithmetic space* condition_op space* (function_call / column_name / quoted_literal / numeric_literal)
+    main_condition        = low_pri_arithmetic space* condition_op space* (function_call / simple_term)
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:126]: Are the changes to this file related with the goal of comparing the SQL string ? If not I would split them into their own PR.

File: snuba/query/timeseries_extension.py [hunk 2]
```diff
@@ -90,12 +90,20 @@ def process_query(
                 BooleanFunctions.AND,
                 binary_condition(
                     ConditionFunctions.GTE,
-                    Column(None, None, self.__timestamp_column),
+                    Column(
+                        f"_snuba_{self.__timestamp_column}",
+                        None,
+                        self.__timestamp_column,
+                    ),
```
Comment (fpacifici) [inline snuba/query/timeseries_extension.py:97]: Same concern. are these changes related with comparing the SQL. 
Also why do we need to add an alias to this column ?

File: tests/conftest.py [hunk 3]
```diff
@@ -104,8 +103,16 @@ def literal(value: Union[str, List[Any]]) -> str:
...
+        aggregations = []
+        for a in legacy.get("aggregations", []):
+            if a[0].endswith(")") and not a[1]:
+                aggregations.append(f"{a[0]} AS {a[2]}")
+            else:
+                agg = func(a)
+                aggregations.append(agg)
+
+        expressions = aggregations + list(map(func, legacy.get("selected_columns", [])))
+        select_clause = f"SELECT {', '.join(expressions)}" if expressions else ""
```
Comment (fpacifici) [inline tests/conftest.py:115]: Did you move the aggregation section in order to ensure expressions are generated in the same order between the legacy query and the snql query.

File: snuba/query/snql/parser.py (hunk 1 above)
Comment (evanh) [reply snuba/query/snql/parser.py:126]: They are related. They stemmed from adding the null_literal.

File: snuba/query/timeseries_extension.py (hunk 2 above)
Comment (evanh) [reply snuba/query/timeseries_extension.py:97]: SnQL mangles the alias for every expression. The regular query parser does too, but these extension conditions are added after the mangling is done. It was easier to add the alias here than to try and skip them in SnQL.

File: tests/conftest.py (hunk 3 above)
Comment (evanh) [reply tests/conftest.py:115]: Yes I did.

File: snuba/query/snql/parser.py [hunk 4]
```diff
@@ -136,7 +141,7 @@
     low_pri_tuple         = low_pri_op space* high_pri_arithmetic
     high_pri_tuple        = high_pri_op space* arithmetic_term
 
-    arithmetic_term       = space* (function_call / numeric_literal / subscriptable / column_name / parenthesized_arithm)
+    arithmetic_term       = space* (function_call / subscriptable / simple_term / parenthesized_arithm)
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:144]: NULL should not be a valid term here. Arithmetic terms are used in conjunction with `low_pri_op` or `hi_pri_op` which are `+ - / * ` and are not supposed to have NULL as an operand.
Also quoted_literal should not be valid here. Any reason to add it ?

File: snuba/query/snql/parser.py [hunk 5]
```diff
@@ -728,8 +745,23 @@ def parse_snql_query_initial(
...
+        idx = e.column()
+        prefix = body[max(0, idx - 1) : idx]
+        suffix = body[idx : (idx + 10)]
+        raise ParsingException(f"Parsing error at '{prefix}{suffix}'")
+    except Exception as e:
+        message = str(e)
+        if "\n" in message:
+            message, _ = message.split("\n", 1)
+        raise ParsingException(message)
+
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:764]: thanks

**Review by fpacifici — APPROVED** [review]
I would suggest, for next time, to split the concerns of the PRs into smaller ones. Specifically the change to the test infrastructure from the bug fixing to the parser that showed up when you changed the test infrastructure. 
Since github does not show the files in the same order you touched them, it is tricky to figure the connection.

File: snuba/query/snql/parser.py (hunk 4 above)
Comment (evanh) [reply snuba/query/snql/parser.py:144]: This is also ultimately used as in the SELECT column as well (selected_expression -> low_pri_arithmetic -> high_pri_arithmetic -> arithmetic_term). The way this is written, it's hard to separate out the different use cases the way you're suggesting. That's why I added it here. I'll add a tech debt story to rearchitect these regexes to allow us to be more specific when validating.

Comment (evanh) [reply snuba/query/snql/parser.py:144]: Also, it is valid to do math with NULL. It just always results in a result of NULL.
